File: src/main/java/me/taubsie/dungeonhub/application/command/commands/EmbedCommand.java
package me.taubsie.dungeonhub.application.command.commands;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import me.taubsie.dungeonhub.application.command.Command;
import me.taubsie.dungeonhub.application.command.CommandParameters;
import me.taubsie.dungeonhub.application.connection.DiscordConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.ContentConnection;
import me.taubsie.dungeonhub.application.enums.EmbedColor;
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionException;
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException;
import me.taubsie.dungeonhub.application.exceptions.InvalidSubCommandException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.common.DungeonHubService;
import org.javacord.api.entity.channel.ChannelType;
import org.javacord.api.entity.channel.ServerChannel;
import org.javacord.api.entity.message.Message;
import org.javacord.api.entity.message.embed.Embed;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.permission.PermissionType;
import org.javacord.api.entity.server.Server;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.*;

import java.awt.*;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletionException;
import java.util.stream.Stream;

@CommandParameters(name = "embed", description = "Makes it possible to manage embeds.", enabledForPermissions =
        {PermissionType.MANAGE_MESSAGES})
public class EmbedCommand extends Command {
    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        SlashCommandInteractionOption firstOption = getOptionAtIndex(0);

        switch (firstOption.getName().toLowerCase()) {
            case "get" -> get(firstOption);
            case "send" -> send(firstOption);
            case "edit" -> edit(firstOption);
            case "add" -> add(firstOption);
            default -> throw new InvalidSubCommandException();
        }
    }

    private void setFields(EmbedBuilder embed, JsonElement value) {
        value.getAsJsonArray().asList().stream()
                .map(JsonElement::getAsJsonObject)
                .forEach(jsonObject -> embed.addField(
                        jsonObject.getAsJsonPrimitive("name").getAsString(),
                        jsonObject.getAsJsonPrimitive("value").getAsString(),
                        jsonObject.getAsJsonPrimitive("inline").getAsBoolean()
                ));
    }

    private void setFooter(EmbedBuilder embed, JsonElement value) {
        JsonObject footer = value.getAsJsonObject();

        String text = footer.getAsJsonPrimitive("text").getAsString();

        if (footer.has("iconUrl")) {
            embed.setFooter(text, footer.getAsJsonPrimitive("iconUrl").getAsString());
        } else {
            embed.setFooter(text);
        }
    }

    private void setThumbnail(EmbedBuilder embed, JsonElement value) {
        JsonObject footer = value.getAsJsonObject();

        embed.setThumbnail(footer.getAsJsonPrimitive("url").getAsString());
    }

    private void applyJson(EmbedBuilder embed, String key, JsonElement value) {
        switch (key) {
            case "title" -> embed.setTitle(value.getAsString());
            case "description" -> embed.setDescription(value.getAsString());
            case "author" -> embed.setAuthor(value.getAsString());
            case "url" -> embed.setUrl(value.getAsString());
            case "color" ->
                    embed.setColor(DungeonHubService.getInstance().getGson().fromJson(value.getAsString(), Color.class));
            case "fields" -> setFields(embed, value);
            case "footer" -> setFooter(embed, value);
            case "timestamp" ->
                    embed.setTimestamp(DungeonHubService.getInstance().getGson().fromJson(value.getAsString(), Instant.class));
            case "thumbnail" -> setThumbnail(embed, value);
        }
    }

    private void add(SlashCommandInteractionOption firstOption) {
        String source = getStringOption(firstOption, "embed");

        List<EmbedBuilder> embeds = new ArrayList<>();

        try {
            JsonElement embedSource = DungeonHubService.getInstance().getGson().fromJson(source, JsonElement.class);

            if (embedSource.isJsonObject()) {
                EmbedBuilder embedBuilder = new EmbedBuilder();

                embedSource.getAsJsonObject()
                        .entrySet()
                        .forEach(entry -> applyJson(embedBuilder, entry.getKey(), entry.getValue()));

                embeds.add(embedBuilder);
            } else if (embedSource.isJsonArray()) {
                embedSource.getAsJsonArray()
                        .forEach(jsonElement -> {
                            if (jsonElement.isJsonObject()) {
                                EmbedBuilder embedBuilder = new EmbedBuilder();

                                jsonElement.getAsJsonObject()
                                        .entrySet()
                                        .forEach(entry -> applyJson(embedBuilder, entry.getKey(), entry.getValue()));

                                embeds.add(embedBuilder);
                            }
                        });
            }
        }
        catch (Exception exception) {
            throw new CommandExecutionException(exception);
        }

        if (embeds.isEmpty()) {
            throw new CommandExecutionException("Please provide any embeds to send.");
        }

        String link = getStringOption(firstOption, "link");

        Message message = DiscordConnection.getInstance().getBot().getMessageByLink(link)
                .orElseThrow(() -> new InvalidOptionException("link"))
                .join();

        if (!message.getAuthor().isYourself()) {
            throw new InvalidOptionException("link", "How should I edit a message that wasn't sent by myself?");
        }

        if (message.getEmbeds().isEmpty()) {
            throw new InvalidOptionException("link", "The given message doesn't have any embeds to edit.");
        }

        List<EmbedBuilder> newEmbeds = new ArrayList<>(message.getEmbeds().stream().map(Embed::toBuilder).toList());

        newEmbeds.addAll(embeds);

        message.edit(newEmbeds);

        respond(ApplicationService.getInstance()
                .getEmbed()
                .setColor(EmbedColor.POSITIVE.getColor())
                .setDescription("Embed(s) added!"));
    }

    private void edit(SlashCommandInteractionOption firstOption) {
        EmbedBuilder embed = new EmbedBuilder();

        try {
            JsonObject embedSource = DungeonHubService.getInstance().getGson().fromJson(getStringOption(firstOption, "embed"), JsonObject.class);

            embedSource.entrySet().forEach(entry -> applyJson(embed, entry.getKey(), entry.getValue()));
        }
        catch (Exception exception) {
            throw new CommandExecutionException(exception);
        }

        int count;
        try {
            count = Math.toIntExact(getLongOption(firstOption, "count"));
        }
        catch (InvalidOptionException ignored) {
            count = 0;
        }

        String link = getStringOption(firstOption, "link");

        Message message = DiscordConnection.getInstance().getBot().getMessageByLink(link)
                .orElseThrow(() -> new InvalidOptionException("link"))
                .join();

        if (!message.getAuthor().isYourself()) {
            throw new InvalidOptionException("link", "How should I edit a message that wasn't sent by myself?");
        }

        if (message.getEmbeds().isEmpty()) {
            throw new InvalidOptionException("link", "The given message doesn't have any embeds to edit.");
        }

        List<EmbedBuilder> embeds = new ArrayList<>(message.getEmbeds().stream().map(Embed::toBuilder).toList());

        if (count >= embeds.size()) {
            throw new InvalidOptionException("link", "The given message doesn't have that many embeds.");
        }

        embeds.set(count, embed);

        message.edit(embeds);

        respond(ApplicationService.getInstance()
                .getEmbed()
                .setColor(EmbedColor.POSITIVE.getColor())
                .setDescription("Embed edited!"));
    }

    private void send(SlashCommandInteractionOption firstOption) {
        ServerChannel channel;

        try {
            channel = getChannelOption(firstOption, "channel");
        }
        catch (InvalidOptionException ignored) {
            channel = getChannel().asServerChannel().orElseThrow();
        }

        String source = getStringOption(firstOption, "embed");

        List<EmbedBuilder> embeds = new ArrayList<>();

        try {
            JsonElement embedSource = DungeonHubService.getInstance().getGson().fromJson(source, JsonElement.class);

            if (embedSource.isJsonObject()) {
                EmbedBuilder embedBuilder = new EmbedBuilder();

                embedSource.getAsJsonObject()
                        .entrySet()
                        .forEach(entry -> applyJson(embedBuilder, entry.getKey(), entry.getValue()));

                embeds.add(embedBuilder);
            } else if (embedSource.isJsonArray()) {
                embedSource.getAsJsonArray()
                        .forEach(jsonElement -> {
                            if (jsonElement.isJsonObject()) {
                                EmbedBuilder embedBuilder = new EmbedBuilder();

                                jsonElement.getAsJsonObject()
                                        .entrySet()
                                        .forEach(entry -> applyJson(embedBuilder, entry.getKey(), entry.getValue()));

                                embeds.add(embedBuilder);
                            }
                        });
            }
        }
        catch (Exception exception) {
            throw new CommandExecutionException(exception);
        }

        if (embeds.isEmpty()) {
            throw new CommandExecutionException("Please provide any embeds to send.");
        }

        try {
            channel.asTextChannel().orElseThrow(() -> new CommandExecutionException("Unknown channel type"))
                    .sendMessage(embeds).join();
        }
        catch (CompletionException ignored) {
            throw new CommandExecutionException("Embed(s) couldn't be sent; please check if all required fields are set.");
        }

        respond(ApplicationService.getInstance()
                .getEmbed()
                .setColor(EmbedColor.POSITIVE.getColor())
                .setDescription("Embed sent!"));
    }

    private void get(SlashCommandInteractionOption firstOption) {
        Server server = getServer();

        String link = getStringOption(firstOption, "link");

        int count;
        try {
            count = Math.toIntExact(getLongOption(firstOption, "count"));
        }
        catch (InvalidOptionException ignored) {
            count = -1;
        }

        Optional<String> type = getOptionalStringOption(firstOption, "type");

        boolean beautiful = type.map(s -> s.equalsIgnoreCase("beautiful")).orElse(false);

        Message message = server.getApi().getMessageByLink(link)
                .orElseThrow(() -> new InvalidOptionException("link"))
                .join();

        if (message.getEmbeds().isEmpty()) {
            throw new InvalidOptionException("link", "The given message doesn't have an embed.");
        }

        if (count != -1 && count >= message.getEmbeds().size()) {
            throw new InvalidOptionException("link", "The given message doesn't have that many embeds.");
        }

        List<Embed> embeds;
        if (count == -1) {
            embeds = message.getEmbeds();
        } else {
            embeds = List.of(message.getEmbeds().get(count));
        }

        EmbedBuilder embedBuilder = ApplicationService.getInstance()
                .getEmbed()
                .setColor(EmbedColor.DEFAULT.getColor());

        if (beautiful && embeds.size() == 1) {
            DungeonHubService.getInstance().getGson().toJsonTree(embeds.get(0))
                    .getAsJsonObject()
                    .entrySet()
                    .forEach(entry -> embedBuilder.addField(entry.getKey(), entry.getValue().toString()));
        } else {
            String embedSource = DungeonHubService.getInstance().getGson().toJson(embeds.size() == 1 ? embeds.get(0) : embeds);

            String description = embedSource;

            if (type.map(s -> s.equalsIgnoreCase("cdn")).orElse(false)) {
                description = ContentConnection.getInstance().uploadFile(embedSource.getBytes(StandardCharsets.UTF_8)).map(s -> ContentConnection.getInstance().getCdnUrl(s).toString()).orElse(embedSource);
            }

            embedBuilder.setDescription(description);
        }

        respond(embedBuilder);
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        SlashCommandOption linkOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("link")
                .setDescription("Please paste the link to the message here.")
                .setRequired(true)
                .build();

        SlashCommandOption typeOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("type")
                .setDescription("Select how you want to get the embed data.")
                .setChoices(Stream.of("beautiful", "source", "cdn").map(s -> new SlashCommandOptionChoiceBuilder().setName(s).setValue(s).build()).toList())
                .setRequired(false)
                .build();

        SlashCommandOption countOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.LONG)
                .setName("count")
                .setDescription("Select which embed you want to get (0-based counting).")
                .setLongMinValue(0)
                .setLongMaxValue(100)
                .setRequired(false)
                .build();

        SlashCommandOption getOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("get")
                .setDescription("Gets the embed data of a message.")
                .setOptions(List.of(linkOption, typeOption, countOption))
                .build();

        SlashCommandOption channelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setName("channel")
                .setDescription("The channel to send the embed into.")
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL, ChannelType.PRIVATE_CHANNEL, ChannelType.SERVER_VOICE_CHANNEL, ChannelType.SERVER_NEWS_CHANNEL, ChannelType.SERVER_PUBLIC_THREAD))
                .setRequired(false)
                .build();

        SlashCommandOption embedOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("embed")
                .setDescription("The embed data to send.")
                .setRequired(true)
                .build();

        SlashCommandOption sendOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("send")
                .setDescription("Sends a custom embed.")
                .setOptions(List.of(embedOption, channelOption))
                .build();

        SlashCommandOption addOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("add")
                .setDescription("Add an embed to a message sent by this bot.")
                .setOptions(List.of(linkOption, embedOption))
                .build();

        SlashCommandOption editOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("edit")
                .setDescription("Edit an embed sent by this bot.")
                .setOptions(List.of(linkOption, embedOption, countOption))
                .build();

        return List.of(getOption, sendOption, addOption, editOption);
    }
}

File: src/main/kotlin/me/taubsie/dungeonhub/application/commands/EmbedCommand.kt
package me.taubsie.dungeonhub.application.commands

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSyntaxException
import dev.kord.common.entity.ChannelType
import dev.kord.common.entity.Permission
import dev.kord.common.entity.Permissions
import dev.kord.core.behavior.channel.asChannelOf
import dev.kord.core.behavior.channel.createMessage
import dev.kord.core.behavior.edit
import dev.kord.core.entity.Message
import dev.kord.core.entity.channel.MessageChannel
import dev.kord.rest.builder.message.EmbedBuilder
import dev.kordex.core.commands.Arguments
import dev.kordex.core.commands.application.slash.converters.impl.optionalStringChoice
import dev.kordex.core.commands.application.slash.publicSubCommand
import dev.kordex.core.commands.converters.impl.optionalChannel
import dev.kordex.core.commands.converters.impl.optionalInt
import dev.kordex.core.commands.converters.impl.string
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.publicSlashCommand
import dev.kordex.core.utils.getJumpUrl
import me.taubsie.dungeonhub.application.config.ConfigProperty
import me.taubsie.dungeonhub.application.connection.*
import me.taubsie.dungeonhub.application.connection.dungeon_hub.ContentConnection
import me.taubsie.dungeonhub.application.enums.EmbedColor
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionException
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException
import me.taubsie.dungeonhub.application.loader.LoadExtension
import me.taubsie.dungeonhub.application.service.ApplicationService
import me.taubsie.dungeonhub.common.DungeonHubService
import java.nio.charset.StandardCharsets
import java.util.function.Consumer

/**
 * This extension provides the ability to manage embeds.
 * It allows you to get the embed data of a message, send a custom embed, add an embed to a message sent by this bot, and edit an embed sent by this bot.
 * The embed data can be provided in a JSON format.
 * The embed data can be displayed as a beautiful embed (all fields of the embed class are mapped to embed fields), as source code, or as a CDN link (which contains the source code).
 * The embed data can be sent to a specific channel and defaults to the channel where the command was executed.
 */
@LoadExtension
class EmbedCommand : Extension() {
    override val name = "embed-command"

    //Suppress "method too complex" because it's pretty understandable and the size is required
    @Suppress("kotlin:S3776")
    override suspend fun setup() {
        publicSlashCommand {
            name = "embed"
            description = "Makes it possible to manage embeds."
            allowInDms = false
            defaultMemberPermissions = Permissions(Permission.ManageMessages)

            publicSubCommand(::GetArguments) {
                name = "get"
                description = "Gets the embed data of a message."

                action {
                    respond {
                        val beautiful = arguments.type.equals("beautiful", ignoreCase = true)

                        val message = arguments.getMessage() ?: throw InvalidOptionException(
                            "link"
                        )

                        if (message.embeds.isEmpty()) {
                            throw InvalidOptionException(
                                "link",
                                "The given message doesn't have an embed."
                            )
                        }

                        val count = arguments.count ?: -1

                        if (count != -1 && count >= message.embeds.size) {
                            throw InvalidOptionException(
                                "link",
                                "The given message doesn't have that many embeds."
                            )
                        }

                        val embeds = if (count == -1) message.embeds else mutableListOf(message.embeds[count])

                        val embedBuilder = ApplicationService.embed
                        embedBuilder.color = EmbedColor.Default.color

                        if (beautiful && embeds.size == 1) {
                            DungeonHubService.getInstance().gson.toJsonTree(embeds[0])
                                .asJsonObject
                                .entrySet()
                                .forEach(Consumer { entry: Map.Entry<String, JsonElement> ->
                                    embedBuilder.field(
                                        entry.key,
                                        false
                                    )
                                    {
                                        entry.value.toString()
                                    }
                                })
                        } else {
                            val embedSource = DungeonHubService.getInstance().gson.toJson(
                                if (embeds.size == 1) embeds[0].toModel()
                                else embeds.map { it.toModel() }
                            )

                            val description =
                                if (embedSource.length >= 4000 || arguments.type.equals("cdn", ignoreCase = true)) {
                                    ContentConnection.getInstance()
                                        .uploadFile(embedSource.toByteArray(StandardCharsets.UTF_8)).map { s: String? ->
                                            ContentConnection.getInstance()
                                                .getCdnUrl(s).toString()
                                        }.orElse(embedSource)
                                } else {
                                    "```\n$embedSource\n```"
                                }

                            embedBuilder.description = description
                        }

                        this.embeds = mutableListOf(embedBuilder)
                    }
                }
            }

            publicSubCommand(::SendArguments) {
                name = "send"
                description = "Sends a custom embed."

                action {
                    respond {
                        val channel = (arguments.channel ?: channel).asChannelOf<MessageChannel>()

                        var source = arguments.embed

                        val cdnPrefix = ConfigProperty.CDN_URL.value
                            ?: throw CommandExecutionException("The CDN isn't set up correctly in the bot's settings, please tell an administrator to correct that.")

                        if (source.startsWith(cdnPrefix)) {
                            source = source.substring(cdnPrefix.length)

                            source =
                                ContentConnection.getInstance()
                                    .downloadFile(source).orElseThrow {
                                        CommandExecutionException(
                                            "Couldn't download the file correctly."
                                        )
                                    }
                        }

                        val embedBuilders: MutableList<EmbedBuilder> = ArrayList()

                        try {
                            val embedSource = DungeonHubService.getInstance().gson.fromJson(
                                source,
                                JsonElement::class.java
                            )

                            if (embedSource.isJsonObject) {
                                val embedBuilder = EmbedBuilder()

                                embedSource.asJsonObject
                                    .entrySet()
                                    .forEach { entry: Map.Entry<String, JsonElement> ->
                                        embedBuilder.applyJson(
                                            entry.key,
                                            entry.value
                                        )
                                    }

                                embedBuilders.add(embedBuilder)
                            } else if (embedSource.isJsonArray) {
                                embedSource.asJsonArray
                                    .forEach { jsonElement: JsonElement ->
                                        if (jsonElement.isJsonObject) {
                                            val embedBuilder = EmbedBuilder()

                                            jsonElement.asJsonObject
                                                .entrySet()
                                                .forEach { entry: Map.Entry<String, JsonElement> ->
                                                    embedBuilder.applyJson(
                                                        entry.key,
                                                        entry.value
                                                    )
                                                }

                                            embedBuilders.add(embedBuilder)
                                        }
                                    }
                            }
                        } catch (jsonSyntaxException: JsonSyntaxException) {
                            throw CommandExecutionException("The embed JSON you entered is invalid formatted.")
                        } catch (exception: Exception) {
                            throw CommandExecutionException(exception)
                        }

                        if (embedBuilders.isEmpty()) {
                            throw CommandExecutionException("Please provide any embeds to send.")
                        }

                        channel.createMessage {
                            embeds = embedBuilders
                        }

                        val embed = ApplicationService.embed
                        embed.color = EmbedColor.Positive.color
                        embed.description = "Embed sent into <#${channel.id}>!"
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::MessageLinkEmbedArguments) {
                name = "add"
                description = "Add an embed to a message sent by this bot."

                action {
                    respond {
                        val embedBuilders: MutableList<EmbedBuilder> = mutableListOf()

                        try {
                            val embedSource = DungeonHubService.getInstance().gson.fromJson(
                                arguments.embed,
                                JsonElement::class.java
                            )

                            if (embedSource.isJsonObject) {
                                val embedBuilder = EmbedBuilder()

                                embedSource.asJsonObject
                                    .entrySet()
                                    .forEach {
                                        embedBuilder.applyJson(
                                            it.key,
                                            it.value
                                        )
                                    }

                                embedBuilders.add(embedBuilder)
                            } else if (embedSource.isJsonArray) {
                                embedSource.asJsonArray
                                    .forEach { jsonElement: JsonElement ->
                                        if (jsonElement.isJsonObject) {
                                            val embedBuilder = EmbedBuilder()

                                            jsonElement.asJsonObject
                                                .entrySet()
                                                .forEach { entry: Map.Entry<String, JsonElement> ->
                                                    embedBuilder.applyJson(
                                                        entry.key,
                                                        entry.value
                                                    )
                                                }

                                            embedBuilders.add(embedBuilder)
                                        }
                                    }
                            }
                        } catch (jsonSyntaxException: JsonSyntaxException) {
                            throw CommandExecutionException("The embed JSON you entered is invalid formatted.")
                        } catch (exception: java.lang.Exception) {
                            throw CommandExecutionException(exception)
                        }

                        if (embedBuilders.isEmpty()) {
                            throw CommandExecutionException("Please provide any embeds to send.")
                        }

                        val message = arguments.getMessage() ?: throw InvalidOptionException(
                            "link"
                        )

                        //TODO how to handle interaction responses?
                        if (message.author?.isSelf() != true) {
                            throw InvalidOptionException(
                                "link",
                                "How should I edit a message that wasn't sent by myself?"
                            )
                        }

                        if (message.embeds.isEmpty()) {
                            throw InvalidOptionException(
                                "link",
                                "The given message doesn't have any embeds to edit."
                            )
                        }

                        message.edit {
                            this@edit.embeds =
                                (message.embeds.toMutableList().map { it.toBuilder() } + embedBuilders).toMutableList()
                        }

                        val embed = ApplicationService.embed
                        embed.color = EmbedColor.Positive.color
                        embed.description = "Embed(s) added to message " + message.getJumpUrl() + "!"
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::EditArguments) {
                name = "edit"
                description = "Edit an embed sent by this bot."

                action {
                    respond {
                        val embed = EmbedBuilder()
                        try {
                            val embedSource: JsonObject = DungeonHubService.getInstance().gson.fromJson(
                                arguments.embed,
                                JsonObject::class.java
                            )

                            embedSource.entrySet().forEach { entry: Map.Entry<String, JsonElement> ->
                                embed.applyJson(
                                    entry.key,
                                    entry.value
                                )
                            }
                        } catch (jsonSyntaxException: JsonSyntaxException) {
                            throw CommandExecutionException("The embed JSON you entered is invalid formatted.")
                        } catch (exception: Exception) {
                            throw CommandExecutionException(exception)
                        }

                        val count = arguments.count ?: 0

                        val message = arguments.getMessage()

                        //TODO how to handle interaction responses?


                        if (message?.author?.isSelf() != true) {
                            throw InvalidOptionException(
                                "link",
                                "How should I edit a message that wasn't sent by myself?"
                            )
                        }

                        if (message.embeds.isEmpty()) {
                            throw InvalidOptionException(
                                "link",
                                "The given message doesn't have any embeds to edit."
                            )
                        }

                        message.edit {
                            val embedBuilders = message.embeds.map { it.toBuilder() }.toMutableList()
                            if (count >= embedBuilders.size) {
                                throw InvalidOptionException(
                                    "link",
                                    "The given message doesn't have that many embeds."
                                )
                            }

                            embedBuilders[count] = embed
                            embeds = embedBuilders
                        }

                        val response = ApplicationService.embed
                        response.color = EmbedColor.Positive.color
                        response.description = "Embed in message ${message.getJumpUrl()} edited!"

                        embeds = mutableListOf(response)
                    }
                }
            }
        }
    }

    open inner class MessageLinkArguments : Arguments() {
        private val messageLink by string {
            name = "link"
            description = "Please paste the link to the message here."
        }

        suspend fun getMessage(): Message? {
            return kord.loadMessageByLink(messageLink)
        }
    }

    inner class GetArguments : MessageLinkArguments() {
        val type by optionalStringChoice {
            name = "type"
            description = "Select how you want to get the embed data."
            choices = mutableMapOf("beautiful" to "beautiful", "source" to "source", "cdn" to "cdn")
        }

        val count by optionalInt {
            name = "count"
            description = "Select which embed you want to get (0-based counting)."
            minValue = 0
            maxValue = 25
        }
    }

    open inner class MessageLinkEmbedArguments : MessageLinkArguments() {
        val embed by string {
            name = "embed"
            description = "The embed data to send."
        }
    }

    inner class SendArguments : Arguments() {
        val embed by string {
            name = "embed"
            description = "The embed data to send."
        }

        val channel by optionalChannel {
            name = "channel"
            description = "The channel to send the embed into."

            requiredChannelTypes = mutableSetOf(
                ChannelType.GuildText,
                ChannelType.GuildVoice,
                ChannelType.GuildNews,
                ChannelType.PublicGuildThread
            )
        }
    }

    inner class EditArguments : MessageLinkEmbedArguments() {
        val count by optionalInt {
            name = "count"
            description = "Select which embed you want to edit (0-based counting)."
            minValue = 0
            maxValue = 25
        }
    }
}